quixstreams/rowconsumer/buffering.py

```diff
@@ -0,0 +1,403 @@
+import logging
+from collections import deque
+from operator import attrgetter
+from typing import Iterable, Optional, Union
+
+from confluent_kafka import TopicPartition
+
+from quixstreams.models import SuccessfulConfluentKafkaMessageProto
+
+logger = logging.getLogger(__name__)
+
+__all__ = ("InternalConsumerBuffer",)
+
+
+_next_timestamp_getter = attrgetter("next_timestamp")
+
+
+class PartitionBuffer:
+    def __init__(
+        self,
+        partition: int,
+        topic: str,
+        max_size: int,
+    ):
+        """
+        A buffer that holds data for a single topic partition.
+
+        :param partition: partition number.
+        :param topic: topic name.
+        :param max_size: the maximum size of the buffer when the buffer is considered full.
+            It is a soft limit, and it may be exceeded in some cases
+        """
+        self.partition = partition
+        self.topic = topic
+        self.max_size = max_size
+        self.high_watermark = -1001
```

Comment:
Please add a comment explaining why it is -1001 as it's not a usual default value. Also why can't the undefined default be just None?
```suggestion
        self.high_watermark: Optional[int] = None
```